### PR TITLE
[DTensor] add _explicit_order_placements util

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150887
* #150862
* #150650
* #150490
* __->__ #150493

The util converts a list of placements in the traditional DTensor format
(e.g. [_StridedShard(0), Shard(0)], where list position is mesh_dim and sharding
is always applied left-to-right (from dim 0 to higher dims))

to a more explicitly ordered format, also replacing '_StridedShard' with
simple 'Shard' placements in the process.
(e.g. the above becomes [(1, Shard(0)), (0, Shard(0)] where the first
item in the tuple is the mesh_dim and the ordering of the tuples is the
sharding order.

This is useful so far as a helper for fixing local shape computation for
strided sharding in the uneven shape case, in the following PR- but may
also be useful more broadly if we can use explicit orderings to simplify
other parts of DTensor logic.

This skips implementing some combinations of _StridedSharding that are
not currently used in the wild today, but could be supported easily.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/tensor/test_utils.py [hunk 1]
```diff
@@ -4,12 +4,15 @@
 
 import torch
 from torch.distributed._tensor import distribute_tensor, DTensor
-from torch.distributed._tensor._utils import compute_local_shape_and_global_offset
+from torch.distributed._tensor._utils import (
```
Comment (XilunWu) [inline test/distributed/tensor/test_utils.py:7]: `_tensor` -> `tensor`

File: test/distributed/tensor/test_utils.py [hunk 2]
```diff
@@ -19,6 +22,70 @@
...
+                "placements": [Shard(0), _StridedShard(0, split_factor=3), Shard(0)],
+                "exception_type": RuntimeError,
+                "exception_text": "Can only convert _StridedShard to ordered Shard if split_factor",
+            },
+            {
+                "mesh_shape": [2, 3, 4],
+                "placements": [
+                    _StridedShard(0, split_factor=6),
+                    _StridedShard(0, split_factor=3),
+                    Shard(0),
```
Comment (XilunWu) [inline test/distributed/tensor/test_utils.py:74]: IIUC, if this test was for case `[(2, Shard(0)), (1, Shard(0)), (0, Shard(0))]`, the corresponding sharding input should've been `[_Strided(0, sf=12), _Strided(0, sf=4), Shard(0)]`.

File: torch/distributed/tensor/_utils.py [hunk 3]
```diff
@@ -15,6 +16,65 @@
 )
 
 
+def _explicit_order_placements(
+    mesh_shape: ShapeType, placements: Sequence[Placement]
```
Comment (XilunWu) [inline torch/distributed/tensor/_utils.py:20]: several util functions including this one can take `DTensorSpec` as input instead of `mesh` + `placements`. I suggest we create a task for this. WDYT?

File: torch/distributed/tensor/_utils.py [hunk 4]
```diff
@@ -15,6 +16,65 @@
...
+
+        else:
+            ordered.append((mesh_dim, p))
+            if isinstance(p, Shard):
+                if p.dim in strided_part_ended_for_dim:
+                    raise NotImplementedError(
+                        f"Strided sharding does not allow Shard() to appear after "
+                        f"the strided part has ended. {p} at mesh dim {mesh_dim} in "
+                        f"{placements} violates this assumption."
+                    )
```
Comment (XilunWu) [inline torch/distributed/tensor/_utils.py:51]: we need a test case for this

**Review by XilunWu — COMMENTED** [review]
This is very helpful in simplifying flexible DTensor sharding order. Left some comments. Besides, it's not quite clear to me how the sharding order is computed. Would like to have an offline discussion.

File: torch/distributed/tensor/_utils.py (hunk 3 above)
Comment (wconstab) [reply torch/distributed/tensor/_utils.py:20]: Ok with me, though one reason I avoided passing mesh and passed mesh shape instead is that I could more easily write a unit test on one process without creating a device mesh. I would like to keep that property somehow.

File: test/distributed/tensor/test_utils.py (hunk 2 above)
Comment (wconstab) [reply test/distributed/tensor/test_utils.py:74]: agreed, that is right. i will correct it.

File: test/distributed/tensor/test_utils.py [hunk 5]
```diff
@@ -19,6 +22,70 @@
...
+            },
+            {
+                "mesh_shape": [2, 2],
+                "placements": [_StridedShard(0, split_factor=2), Shard(0)],
+                "ordered": [(1, Shard(0)), (0, Shard(0))],
+            },
+            {
+                "mesh_shape": [2, 3, 4],
+                "placements": [Shard(0), _StridedShard(0, split_factor=4), Shard(0)],
+                "ordered": [(0, Shard(0)), (2, Shard(0)), (1, Shard(0))],
```
Comment (wconstab) [inline test/distributed/tensor/test_utils.py:47]: @XilunWu now i think this is wrong.  since `(0, Shard(0))` and  `(2, Shard(0))` both happen before the _StridedShard(0), I think the split_factor should be 2 * 4 = 8?  When writing this test I was ignoring that left-hand Shard(0) when considering the strided sharding.

Comment (wconstab) [reply test/distributed/tensor/test_utils.py:47]: ok, yea, i agree.  
Wrong understanding:  Split factor accounts for all of the 'earlier' shardings.  (e.g. dim 0 and then dim 2)
Correct understanding: SF accounts for only the 'earlier' shardings to the right of the SS, since those are the ones we are 'striding' over. (e.g. just dim 2)

File: torch/distributed/tensor/_utils.py [hunk 6]
```diff
@@ -15,6 +16,65 @@
...
+                    while len(strided_placements) > 0:
+                        # We can process multiple strided shardings in reverse-order
+                        # (e.g. [_StridedShard(0, split_factor=4), _StridedShard(0, split_factor=2), Shard(0)])
+                        # TODO- validate this logic and enable it (mainly, validate aggregate_size part)
+                        if len(strided_placements) > 1:
+                            raise NotImplementedError(
+                                "NYI nested strided sharding conversion to ordered"
+                            )
+
+                        strided_dim, strided = strided_placements.pop()
```
Comment (wanchaol) [inline torch/distributed/tensor/_utils.py:66]:
```suggestion
                        strided_mesh_dim, strided = strided_placements.pop()
```
Let's explicitly name it with mesh_dim

File: test/distributed/tensor/test_utils.py [hunk 7]
```diff
@@ -19,6 +22,80 @@
...
+                "placements": [Replicate()],
+                "ordered": [(0, Replicate())],
+            },
+            {
+                "mesh_shape": [2, 2],
+                "placements": [Shard(0)],
+                "ordered": [(0, Shard(0))],
+            },
+            {
+                "mesh_shape": [2, 2],
```
Comment (wanchaol) [inline test/distributed/tensor/test_utils.py:40]: can you make the `mesh_shape` have different mesh dim size instead, i.e. [2, 4] or [4, 2] to test the correctness

File: test/distributed/tensor/test_utils.py [hunk 8]
```diff
@@ -19,6 +22,80 @@
 c10d_functional = torch.ops.c10d_functional
 
 
+class LocalTest(TestCase):
+    def test_explicit_order_placements(self):
+        # mesh_shape: ShapeType, placements: Sequence[Placement]
+        test_cases = [
+            {
+                "mesh_shape": [2, 2],
+                "placements": [Replicate()],
```
Comment (wanchaol) [inline test/distributed/tensor/test_utils.py:31]: hmmm aren't this need to error out? The rank of mesh_shape and placements should be the same

File: test/distributed/tensor/test_utils.py [hunk 9]
```diff
@@ -19,6 +22,80 @@
...
+                "ordered": [(0, Shard(0))],
+            },
+            {
+                "mesh_shape": [2, 2],
+                "placements": [_StridedShard(0, split_factor=2), Shard(0)],
+                "ordered": [(1, Shard(0)), (0, Shard(0))],
+            },
+            {
+                "mesh_shape": [2, 3, 4],
+                "placements": [Shard(0), _StridedShard(0, split_factor=4), Shard(0)],
```
Comment (wanchaol) [inline test/distributed/tensor/test_utils.py:46]: should add a test have SS in the leftmost of the placements

File: torch/distributed/tensor/_utils.py [hunk 10]
```diff
@@ -15,6 +16,65 @@
...
+                        if not strided.split_factor == aggregate_size:
+                            raise RuntimeError(
+                                f"Can only convert _StridedShard to ordered Shard if split_factor({strided.split_factor})"
+                                f" == aggregate mesh size ({aggregate_size})"
+                            )
+                        aggregate_size *= mesh_shape[strided_dim]
```
Comment (wanchaol) [inline torch/distributed/tensor/_utils.py:72]: It looks to me that this might not be right? It probably only works on strided shard be the second right most mesh dim

The aggregated size should not be computed with the strided_mesh_dim, but the mesh_dims other than the strided_mesh_dim

**Review by wanchaol — CHANGES_REQUESTED** [review]
I think the calculation on aggregated size is wrong, we need to create more test cases to verify. Please see inlined comments

File: torch/distributed/tensor/_utils.py [hunk 11]
```diff
@@ -15,6 +16,65 @@
...
+
+    Returns a list of (mesh_dim, placement) tuples where the list order is the sharding order.
+
+    ex.
+    [Shard(0), _StridedShard(0, split_factor=2), Shard(0)] ->
+    [(0, Shard(0)), (2, Shard(0)), (1, Shard(0))]
+
+    """
+    if not any(isinstance(p, _StridedShard) for p in placements):
+        return list(enumerate(placements))
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/tensor/_utils.py:33]: It should allow shard/replicate as input, but I think it only needs to do special handling for the shard placements.

File: torch/distributed/tensor/_utils.py [hunk 12]
```diff
@@ -15,6 +16,65 @@
...
+            # validate the stride is the correct multiple of the meshdim and the earlier shard
+            deferred_strided_placements[p.dim].append((mesh_dim, p))
+
+        else:
+            ordered.append((mesh_dim, p))
+            if isinstance(p, Shard):
+                if p.dim in strided_part_ended_for_dim:
+                    raise NotImplementedError(
+                        f"Strided sharding does not allow Shard() to appear after "
+                        f"the strided part has ended. {p} at mesh dim {mesh_dim} in "
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/tensor/_utils.py:49]: This is actually just a  copy / move of original error message in the existing code.  I could rephrase it but I kept the same words because of matching existing unit test.

What it means is, 

[StridedShard, Shard, Shard] <-  the dim2 Shard in this case is "invalid" because the dim1 shard 'ends' the strided shard region, and there can be no more shards after that.  That's basically a design assumption of 'strided shard' as far as i understood it- cc @XilunWu.  This feels pretty limited to me, hence why I think its cleaner to just remove strided shard and use 'ordering + shard' instead. 

otoh,
[stridedshard, stridedshard, shard] <- this case is supposed to be valid, in the 'design' of strided sharding.  But iiuc this is also not  really implemented (in terms of 'to_replicate_tensor' functionality).  is this correct @XilunWu ?

File: test/distributed/tensor/test_utils.py (hunk 8 above)
Comment (wconstab) [reply test/distributed/tensor/test_utils.py:31]: hmm. yea, i probably need to change the function to be more careful here. and the test case should be 2 replicates, ill change that and also test the one replicate one throws an error.

File: torch/distributed/tensor/_utils.py (hunk 10 above)
Comment (wconstab) [reply torch/distributed/tensor/_utils.py:72]: ill have to think about this, and try to come up with a test case that exposes it if its a bug.

File: test/distributed/tensor/test_utils.py (hunk 9 above)
Comment (wconstab) [reply test/distributed/tensor/test_utils.py:46]: oh, i actually already do.  I have the one right above this one with (SS, S), and below I have the error case for NYI for (SS, SS, S).  I think i can just finish that implementation now though.

File: torch/distributed/tensor/_utils.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/tensor/_utils.py:33]: a couple of follow up thoughts, (1) maybe your point was i should assert that all the placements are in (Shard, Replicate) before taking this path?  But also (2) i think  i can just delete this 'fast path' and rely on the below logic to cover all cases, seems cleaner to just have one path.

File: torch/distributed/tensor/_utils.py (hunk 10 above)
Comment (wconstab) [reply torch/distributed/tensor/_utils.py:72]: I think this is correct.  And I added another test case for (SS_a, SS_b, S) which passes with this.  The reasoning would be

First iter:
- deferred_strided_placements: [SS_a, SS_b]
- process S
- set aggregate_size = size(mesh[2])

Second iter:
- process SS_b, expecting split-factor == size(mesh[2])
- set aggregate_size = size(mesh[2]) * size(mesh[1])

Third iter:
- process SS_a, expecting split-factor == size(mesh[2]) * size(mesh[1])
- set aggregate_size = size(mesh[2]) size(mesh[1]) * size(mesh[0])   - but this value will be ignored

Finished

**Review by wanchaol — APPROVED** [review]
lgtm

File: torch/distributed/tensor/_utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/tensor/_utils.py:51]: I think we do support `[stridedshard, stridedshard, shard]`. What we don't support is `[StridedShard, Shard, StridedShard, Shard]`. 

Note that this is not a valid one (i.e. for any sharding dimension, there should be only one "strided shard region"). But an equivalent one is `[StridedShard, StridedShard, StridedShard, Shard]`, with non-decreasing `split_factor`. I think it's correctly handled in your change.

**Review by XilunWu — APPROVED** [review]
